engine-types/src/types/wei.rs

```diff
@@ -0,0 +1,231 @@
+    }
+
+    pub fn checked_add(self, rhs: Self) -> Option<Self> {
+        self.0.checked_add(rhs.0).map(Self)
+    }
+
+    /// Try convert U256 to u128 with checking overflow.
+    /// NOTICE: Error can contain only overflow
+    pub fn try_into_u128(self) -> Result<u128, error::BalanceOverflowError> {
+        use crate::TryInto;
+        self.0.try_into().map_err(|_| error::BalanceOverflowError)
+    }
+}
+
+impl BorshSerialize for Wei {
+    fn serialize<W: io::Write>(&self, writer: &mut W) -> io::Result<()> {
+        let mut buf = [0u8; 32];
+        self.0.to_big_endian(&mut buf);
+        writer.write_all(&buf)
+    }
+}
+
+impl BorshDeserialize for Wei {
+    fn deserialize(buf: &mut &[u8]) -> io::Result<Self> {
+        Ok(Wei::new(U256::from_big_endian(buf)))
+    }
+
+    fn try_from_slice(v: &[u8]) -> io::Result<Self> {
+        let mut v_mut = v;
+        Self::deserialize(&mut v_mut)
+    }
+}
+
+impl Display for Wei {
+    fn fmt(&self, f: &mut Formatter<'_>) -> crate::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Add<Self> for Wei {
+    type Output = Wei;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Self(self.0 + rhs.0)
+    }
+}
+
+impl Sub<Self> for Wei {
+    type Output = Wei;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        Self(self.0 - rhs.0)
+    }
+}
+
+/// Type casting from Wei compatible Borsh-encoded raw value into the Wei value, to attach an ETH balance to the transaction
+impl From<WeiU256> for Wei {
+    fn from(value: WeiU256) -> Self {
+        Wei(U256::from_big_endian(&value))
+    }
+}
+
+impl From<Fee> for Wei {
+    fn from(value: Fee) -> Self {
+        Wei(U256::from(value.as_u128()))
+    }
+}
+
+impl From<Balance> for Wei {
+    fn from(value: Balance) -> Self {
+        Wei(U256::from(value.as_u128()))
+    }
+}
+
+impl From<NEP141Wei> for Wei {
+    fn from(value: NEP141Wei) -> Self {
+        Wei(U256::from(value.as_u128()))
+    }
+}
+
+#[allow(dead_code)]
+pub fn u256_to_arr(value: &U256) -> [u8; 32] {
+    let mut result = [0u8; 32];
+    value.to_big_endian(&mut result);
+    result
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_wei_from_eth() {
+        let eth_amount: u64 = rand::random();
+        let wei_amount = U256::from(eth_amount) * U256::from(10).pow(18.into());
+        assert_eq!(Wei::from_eth(eth_amount.into()), Some(Wei::new(wei_amount)));
+    }
+
+    #[test]
+    fn test_wei_from_u64() {
+        let x: u64 = rand::random();
+        assert_eq!(Wei::new_u64(x).raw().as_u64(), x);
+    }
+
+    #[test]
+    fn test_wei_serializer() {
+        // borsh serialize
+        let serialized = Wei::new(U256::from(100)).try_to_vec().unwrap();
+        assert_eq!(serialized.len(), 32);
+
+        let balance = Wei::try_from_slice(&serialized).unwrap();
+        assert_eq!(balance, Wei::new(U256::from(100)));
+    }
+
+    #[test]
+    fn test_wei_serialized_len() {
+        let mut serialized = [0u8; 30];
```

Comment:
A little awkward here that the slice is not 32 bytes, but the test still passes. Makes me wonder if we should add a check the to the deserialize implementation that the slice is 32 bytes.

On the one hand, of course a number represented by fewer than 32 bytes could be cast into a U256, but on the other hand it means we could silently ignore some bugs if we know all valid serializations of this type produce 32 bytes.